src/index.js

```diff
@@ -16,102 +14,148 @@ function makeDnsPromise(host) {
   });
 }
 
-function setupTimeoutCapture(config, report, context = {}) {
+function setupTimeoutCapture(wrapperInstance) {
+  const { modifiedContext, sendReport, config } = wrapperInstance;
   var endTime = 599900; /* Maximum execution: 100ms short of 5 minutes */
   if (config.timeoutWindow < 1) {
     return undefined;
   }
 
-  if (config.timeoutWindow > 0 && context.getRemainingTimeInMillis) {
+  if (config.timeoutWindow > 0 && modifiedContext.getRemainingTimeInMillis) {
     endTime = Math.max(
       0,
-      context.getRemainingTimeInMillis() - config.timeoutWindow
+      modifiedContext.getRemainingTimeInMillis() - config.timeoutWindow
     );
   }
 
   return setTimeout(() => {
-    report.send(new Error('Timeout Exceeded.'), function noop() {});
+    sendReport(new Error('Timeout Exceeded.'));
   }, endTime);
 }
 
-module.exports = options => {
-  const fn = func => {
-    var dnsPromise = undefined;
-    fn.metricsQueue = [];
-    const config = setConfig(options);
+class IOpipeWrapperClass {
+  constructor(dnsPromise, config, userFunc) {
+    this.config = config;
+    this.metrics = [];
+    this.userFunc = userFunc;
 
-    if (!config.clientId) {
-      // No-op if user doesn't set an IOpipe token.
-      return func;
+    // assign a new dnsPromise if it's not a coldstart because dns could have changed
+    if (!globals.COLDSTART) {
```

Comment:
this will never get executed now. The constructor is only happening on coldstarts anyway, now.

Perhaps just do the "when coldstart" here and then the "when not coldstart" block under invoke()?